exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/http/logs/OtlpHttpLogRecordExporterBuilder.java

```diff
@@ -210,8 +210,13 @@ public OtlpHttpLogRecordExporterBuilder setMeterProvider(
     return this;
   }
 
-  /** Set the {@link MemoryMode}. */
-  OtlpHttpLogRecordExporterBuilder setMemoryMode(MemoryMode memoryMode) {
+  /**
+   * Set the {@link MemoryMode}. If unset, defaults to {@link #DEFAULT_MEMORY_MODE}.
+   *
+   * <p>>When memory mode is {@link MemoryMode#REUSABLE_DATA}, serialization is optimized to reduce
+   * memory allocation.
+   */
+  public OtlpHttpLogRecordExporterBuilder setMemoryMode(MemoryMode memoryMode) {
```

Comment:
Do I understand correctly that this is part of public api of a stable module and once we have added this method we can't easily get rid of it? If so then perhaps it would be better to not tie this to implementation details like immutable or reusable data, but rather something abstract like minimize allocations and maximize throughput (I guess this would only make sense if the one that allocates more has a bit better performance). That way the behavior of this method could more easily accommodate future changes which could, for example, include deleting one of the memory modes.

Reply:
> Do I understand correctly that this is part of public api of a stable module and once we have added this method we can't easily get rid of it?

We can't get rid of it without a major version bump, which we don't plan on doing anytime soon.

> perhaps it would be better to not tie this to implementation details like immutable or reusable data, but rather something abstract like minimize allocations and maximize throughput (I guess this would only make sense if the one that allocates more has a bit better performance). That way the behavior of this method could more easily accommodate future changes which could, for example, include deleting one of the memory modes.

The MemoryMode enum is also part of the stable API at this point so we can't delete the memory modes. In hindsight it might have been preferable to choose a name like `MemoryMode.LOW_ALLOCATION` instead `MemoryMode.REUSABLE_DATA`. LOW_ALLOCATION describes the intended outcome (more user facing) where REUSABLE_DATA describes what is happening under the covers. The flip side of this is that REUSABLE_DATA communicates some important information to MetricReader / MetricExporter implementations: we're going to reuse MetricData classes so they won't function right after the CompletableResultCode from MetricExporter.export() resolves. So there's benefits to naming for the outcome and also how the outcome is accomplished.

REUSABLE_DATA doesn't describes what's happening with the serializers as well as it describes what's happening with the metrics SDK where it was originally introduced, but its still somewhat accurate. One thing that sticks out is that with the metrics SDK, implementers of MetricReader / MetricExporter have to be aware of the semantics of REUSABLE_DATA. With serializers, there's no impact to user semantics. The only indication a user can see that something has changed is a shift in the CPU / memory behavior.

Overall, I think its preferable to have one memory mode configuration concept, even if the words we chose for that concept (IMMUTABLE_DATA and REUSABLE_DATA) don't perfectly describe all the places we make that configurable.

Reply:
> The flip side of this is that REUSABLE_DATA communicates some important information to MetricReader / MetricExporter implementations: we're going to reuse MetricData classes so they won't function right after the CompletableResultCode from MetricExporter.export() resolves

👍